### PR TITLE
feat: add iframe cache instantiation and logic

### PR DESCRIPTION
**Pull request by jshrt**
- [ ] Change the url of `3box-shared-cache` to point to `#master`

### CODE REVIEW

File: src/replicator.js [hunk 1]
```diff
@@ -93,15 +95,27 @@ class Replicator {
...
+      const iframe = document.querySelector('iframe')
+      const postMessage = iframe.contentWindow.postMessage.bind(iframe.contentWindow)
+
+      const cacheStorageProxy = await SharedCache.createOrbitStorageProxy(cachePath, { postMessage })
+
+      cache = new OdbCache(cacheStorageProxy)
+    } else {
+      const cacheStorage = await levelDown.createStore(cachePath)
+      cache = new OdbCache(cacheStorage)
+    }
```
Comment (oed) [inline src/replicator.js]: Doesn't seem right to have the document query selector in this class. Should be abstracted away to happen only in 3box.js maybe? The `cacheStorageProxy` could be sent as an option to the Replicator class perhaps.

File: src/3box.js [hunk 2]
```diff
@@ -529,16 +515,51 @@ class Box extends BoxApi {
   }
 }
 
-function initIPFSRepo () {
+function createIframeCache () {
+  const iframe = document.createElement('iframe')
+
+  iframe.src = process.env.CACHE_IFRAME_URL
```
Comment (oed) [inline src/3box.js]: The source should be hardcoded in the `config.js` module, with an option to change it.

File: example/index.js [hunk 3]
```diff
@@ -3,7 +3,7 @@ const syncComplete = (res) => {
   updateProfileData(window.box)
 }
 
-Box.create(window.ethereum).then(box => {
+Box.create(window.ethereum, { iframeCache: true }).then(box => {
```
Comment (oed) [inline example/index.js]: Using this should be default. There should be an option to disable it.

File: webpack.config.js [hunk 4]
```diff
@@ -29,6 +30,11 @@ module.exports = {
       }
     ]
   },
+  plugins: [
+    new webpack.EnvironmentPlugin([
+      'CACHE_IFRAME_URL',
+    ])
+  ],
```
Comment (oed) [inline webpack.config.js]: This isn't needed if we set it it `config.js`

File: src/replicator.js (hunk 1 above)
Comment (jshrt) [reply src/replicator.js]: :thinking: I will look into it. Passing the cacheStorageProxy as a dependency is good idea as well.

File: src/3box.js (hunk 2 above)
Comment (jshrt) [reply src/3box.js]: I did not catch `config.js`, thanks for pointing that out!

File: webpack.config.js (hunk 4 above)
Comment (jshrt) [reply webpack.config.js]: Noted

File: example/index.js (hunk 3 above)
Comment (jshrt) [reply example/index.js]: Gotcha

File: src/3box.js [hunk 5]
```diff
@@ -18,6 +18,7 @@ const LevelStore = require('datastore-level')
 const didJWT = require('did-jwt')
 
 const SharedCache = require('3box-shared-cache')
+const { iframe_cache_url } = require('./config') // eslint-disable-line
```
Comment (oed) [inline src/3box.js]: You don't need to import this since it's already imported on line 14 :)

File: src/3box.js [hunk 6]
```diff
@@ -529,16 +522,68 @@ class Box extends BoxApi {
   }
 }
 
-function initIPFSRepo () {
+function createIframeCache () {
```
Comment (oed) [inline src/3box.js]: This function should probably be called on pageload. Otherwise loading the iframe will slow things down.

Comment (oed) [reply src/3box.js]: cc @zachferland for input.

File: src/config.js [hunk 7]
```diff
@@ -10,8 +9,7 @@ module.exports = {
   address_server_url: process.env.ADDRESS_SERVER_URL || 'https://beta.3box.io/address-server',
   pinning_node: process.env.PINNING_NODE || '/dnsaddr/ipfs.3box.io/tcp/443/wss/ipfs/QmZvxEpiVNjmNbEKyQGvFzAY1BwmGuuvdUTmcTstQPhyVC',
   pinning_room: process.env.PINNING_ROOM || '3box-pinning',
-  iframe_store_version: process.env.IFRAME_STORE_VERSION || IFRAME_STORE_VERSION,
-  iframe_store_url: process.env.IFRAME_STORE_URL || `https://iframe.3box.io/${IFRAME_STORE_VERSION}/iframe.html`,
+  iframe_cache_url: process.env.IFRAME_CACHE_URL || 'https://iframe.3box.io/iframe.html',
```
Comment (oed) [inline src/config.js]: Can we change this to be `https://cache.3box.io`?

File: webpack.config.js [hunk 8]
```diff
@@ -29,6 +30,11 @@ module.exports = {
       }
     ]
   },
+  plugins: [
+    new webpack.EnvironmentPlugin([
+      'IFRAME_CACHE_URL',
+    ])
+  ],
```
Comment (oed) [inline webpack.config.js]: Is this really needed?

File: src/config.js (hunk 7 above)
Comment (zachferland) [reply src/config.js]: yeah

File: src/3box.js (hunk 6 above)
Comment (zachferland) [reply src/3box.js]: yeah, if in browser env, should optimistically load on pageload, and then if any configuration options passed later on function call differ, reload or ignore (mostly want to optimize default then can add different configuration opts/paths if people need it) 

think we could look to move most of this code out of 3boxjs as well, in attempt to make changes as minimal as possible here and for all iframe code to be more contained. I recently did took that path for idw iframe, after doing it more like this last time, before release of both we can try to share some of the same patterns. Lib could export iframe cache service, with similar interface as well (create, func to get orbit proxy, and func to get ipfs proxy). have to see if anything makes that difficult though

File: src/config.js (hunk 7 above)
Comment (jshrt) [reply src/config.js:12]: yup

File: src/3box.js (hunk 5 above)
Comment (jshrt) [reply src/3box.js:21]: Thanks for the catch

**Comment by zachferland** [comment]
Replaced by - https://github.com/3box/3box-js/pull/782

**Pull request closed without merging**